### PR TITLE
Rosetta: include nonce (and memo) in zkapp-command metadata

### PR DESCRIPTION
**Pull request by dkijania**
`Zkapp_command_info.T.to_transaction` was hard-coding `metadata = None`, so the `/block` endpoint returned zk-app transactions without a `nonce` even though `Zkapp_command_info.t` carries it. OKX flagged this while integrating the exchange.

Mirror the memo-decoding pattern from `User_command_info.to_transaction` but always emit the nonce: metadata is now `Some (`Assoc [...])` with `nonce` always present and `memo` appended when the base58-check memo decodes to a non-empty string.

Adds five inline tests covering: None memo, unparseable memo, valid non-empty memo, empty-decoded memo, and a sweep over all dummies asserting metadata is never None (regression pin for the reported bug).

### CODE REVIEW

File: src/app/rosetta/lib/commands_common.ml [hunk 1]
```diff
@@ -331,10 +331,24 @@ module Zkapp_command_info = struct
...
+      let nonce = ("nonce", `Int (Unsigned.UInt32.to_int cmd.nonce)) in
+      let memo_field =
+        Option.bind cmd.memo ~f:(fun base58_check ->
+            try
+              let memo =
+                let open Mina_base.Signed_command_memo in
+                base58_check |> of_base58_check_exn |> to_string_hum
+              in
+              if String.is_empty memo then None else Some ("memo", `String memo)
+            with _ -> None )
```
Comment (glyh) [inline src/app/rosetta/lib/commands_common.ml]: In this exception case, it seems to be base58 check failed? Could we add some loggings here?

**Comment by glyh** [comment]
Looks okay, but is there a spec for Rosetta I can check?

**Comment by dkijania** [comment]
@glyh To be honest i didn't found any specification as zkapp transaction is not standard for rosetta. My approach was to mimic what we did for rosetta for user transaction

**Comment by glyh** [comment]
So this is not really a bug fix, but feature request.

File: src/app/rosetta/lib/commands_common.ml [hunk 2]
```diff
@@ -361,4 +386,86 @@ module Zkapp_command_info = struct
...
+      module T_result = T (Result)
+
+      let metadata_fields cmd =
+        match T_result.to_transaction cmd with
+        | Ok { metadata = Some (`Assoc fields); _ } ->
+            fields
+        | Ok _ ->
+            failwith "expected metadata to be Some (`Assoc _)"
+        | Error _ ->
+            failwith "to_transaction returned Error for a valid command"
```
Comment (glyh) [inline src/app/rosetta/lib/commands_common.ml]: May worth to raise the exception message here.

File: src/app/rosetta/lib/commands_common.ml [hunk 3]
```diff
@@ -361,4 +386,86 @@ module Zkapp_command_info = struct
...
+        let fields = metadata_fields cmd in
+        ( match find_field fields "nonce" with
+        | Some (`Int n) ->
+            [%test_eq: int] n (Unsigned.UInt32.to_int cmd.nonce)
+        | _ ->
+            failwith "expected nonce field to be Some (`Int _)" ) ;
+        [%test_eq: bool] (Option.is_some (find_field fields "memo")) false
+
+      let%test_unit "memo included when valid base58 check string" =
+        let memo_text = "rosetta-okx-test" in
```
Comment (glyh) [inline src/app/rosetta/lib/commands_common.ml]: Maybe make this a property based test?

File: src/app/rosetta/lib/commands_common.ml [hunk 4]
```diff
@@ -361,4 +386,86 @@ module Zkapp_command_info = struct
       ; account_updates = Zkapp_account_update_info.dummies
       }
     ]
+
+  let%test_module "Zkapp_command_info.to_transaction metadata" =
```
Comment (glyh) [inline src/app/rosetta/lib/commands_common.ml:390]: Maybe we can have more coverage by converting commands generated by our command gen instead of using the dummies that we write by hand above.

**Review by glyh — COMMENTED** [review]
Left some comments

File: src/app/rosetta/lib/commands_common.ml (hunk 2 above)
Comment (dkijania) [reply src/app/rosetta/lib/commands_common.ml]: Done — failwith messages now include the actual error/yojson value (using Errors.show for Error _ and Yojson.Safe.to_string for unexpected metadata shapes). See 10b76e1eee.

File: src/app/rosetta/lib/commands_common.ml (hunk 3 above)
Comment (dkijania) [reply src/app/rosetta/lib/commands_common.ml]: Done — added a property-based test using Signed_command_memo.gen plus arbitrary fuzzed strings (and None). It runs 200 trials and asserts: to_transaction always succeeds, metadata is always Some (`Assoc _), nonce always matches the input, and memo is present iff the input decodes to a non-empty plaintext. See 10b76e1eee. The concrete dummy-based tests are kept since they pin specific edge cases (empty decoded memo, a known unparseable input) that would be cumbersome to assert from a generator.

File: src/app/rosetta/lib/commands_common.ml (hunk 4 above)
Comment (dkijania) [reply src/app/rosetta/lib/commands_common.ml:390]: Good idea, but I held off on plugging the actual command generators (e.g., Mina_base.Zkapp_command.gen) into Zkapp_command_info dummies for this PR — converting between Mina_base.Zkapp_command.t and the rosetta-side Zkapp_command_info.t is non-trivial and would expand 3.4.0-bound scope. The PBT in 10b76e1eee fuzzes the bug-relevant axis (memo decoding) directly via Signed_command_memo.gen, which covers the regression. Happy to file a follow-up to wire the generators end-to-end if you want broader coverage.

File: src/app/rosetta/lib/commands_common.ml [hunk 5]
```diff
@@ -465,7 +468,62 @@ module Zkapp_command_info = struct
...
+
+      (* Property-based test: for any random valid memo (generated by
+         Signed_command_memo.gen) and for arbitrary fuzzed memo strings
+         (potentially invalid base58-check), to_transaction must:
+           - always return Ok
+           - always emit Some (`Assoc _) metadata
+           - always include a nonce equal to the input nonce
+           - include a memo iff the input decodes to a non-empty plaintext *)
+      let%test_unit "PBT: metadata invariants over random memos" =
+        let base = List.nth_exn dummies 0 in
```
Comment (glyh) [inline src/app/rosetta/lib/commands_common.ml]: Choose base at random?

File: src/app/rosetta/lib/commands_common.ml [hunk 6]
```diff
@@ -465,7 +468,62 @@ module Zkapp_command_info = struct
...
+        let memo_input_gen =
+          let open Quickcheck.Generator.Let_syntax in
+          let valid_memo =
+            let%map m = Mina_base.Signed_command_memo.gen in
+            `Valid (Mina_base.Signed_command_memo.to_base58_check m)
+          in
+          let arbitrary_string =
+            let%map s = String.quickcheck_generator in
+            `Arbitrary s
```
Comment (glyh) [inline src/app/rosetta/lib/commands_common.ml]: what's the point distinguishing Valid and Aribitrary?

File: src/app/rosetta/lib/commands_common.ml [hunk 7]
```diff
@@ -465,7 +468,62 @@ module Zkapp_command_info = struct
...
+            let expect_memo_field =
+              match memo_decoded with
+              | Some m when not (String.is_empty m) ->
+                  true
+              | _ ->
+                  false
+            in
+            [%test_eq: bool]
+              (Option.is_some (find_field fields "memo"))
+              expect_memo_field )
```
Comment (glyh) [inline src/app/rosetta/lib/commands_common.ml:543]: I'd consider breaking this into 2 tests where one only consists of valid memo and another not.

**Review by glyh — COMMENTED** [review]
Left some comments

**Comment by dkijania** [comment]
!ci-build-me

File: src/app/rosetta/lib/commands_common.ml (hunk 6 above)
Comment (dkijania) [reply src/app/rosetta/lib/commands_common.ml]: Good point. Reverting

**Review by glyh — APPROVED** [review]

**Comment by dkijania** [comment]
!ci-build-me
